### PR TITLE
Ticket 3

### LINKED ISSUE #4: Just for you to know
https://github.com/Ledostuff/calories-counter/blob/71987d4faa86ffbf8f87d8176cdd9766855cdf84/src/main/scala/ru/ledostuff/calories/domain/database/ProductService.scala#L19

прямо с пакетом расширений для функтора, кроме `as`, приходит ещё функция `void` которая делает как раз `.as(())`

### PR DESCRIPTION
**Pull request by Ledostuff**
Prettifying IO operations
Give up using Seq
Use void instead of as(())
@Phill101, another one pull request

### CODE REVIEW

File: src/main/scala/ru/ledostuff/calories/infrastructure/services/ProductCaloriesService.scala [hunk 1]
```diff
@@ -42,23 +42,23 @@ class ProductCaloriesService[F[_]: Sync](productService: ProductService[F],
...
+          translateService.translateProductName(name).value
         }, TranslationNotFound)
 
-        savedProduct          <- EitherT.liftF[F, ErrorResponse, Product](productService.createProduct(Product(name, translatedProductInfo.translatedNames.toSeq)))
+        savedProduct          <- EitherT.liftF[F, ErrorResponse, Product](productService.createProduct(Product(name, translatedProductInfo.translatedNames.toList)))
 
         caloriesInfo          <- {
-          val engNamesRequests = savedProduct.engNames.toList.map { engName =>
+          val engNamesRequests = savedProduct.engNames.map { engName =>
             (EitherT.liftF[F, ErrorResponse, Unit](log.info(s"search calories for name $engName")) *>
```
Comment (Phill101) [inline src/main/scala/ru/ledostuff/calories/infrastructure/services/ProductCaloriesService.scala]: I wanted you to make it everywhere :) Perhaps you missed it

File: src/main/scala/ru/ledostuff/calories/infrastructure/services/ProductCaloriesService.scala [hunk 2]
```diff
@@ -42,23 +42,23 @@ class ProductCaloriesService[F[_]: Sync](productService: ProductService[F],
...
               EitherT.fromOptionF[F, ErrorResponse, ProductCalories](productCaloriesService.getCaloriesByProductName(engName).value, ProductCaloriesNotFound)).value
           }
           EitherT.liftF(engNamesRequests.sequence.map(l => l.collect{case Right(e) => e})).ensure(ProductCaloriesNotFound)(_.nonEmpty)
```
Comment (Phill101) [inline src/main/scala/ru/ledostuff/calories/infrastructure/services/ProductCaloriesService.scala]: Why didn't you use `traverse` instead of `sequence`?

**Comment by Phill101** [comment]
looks good

File: src/main/scala/ru/ledostuff/calories/infrastructure/services/ProductCaloriesService.scala (hunk 1 above)
Comment (Ledostuff) [reply src/main/scala/ru/ledostuff/calories/infrastructure/services/ProductCaloriesService.scala]: Nonono :) I am calling log before each request to the calories service

File: src/main/scala/ru/ledostuff/calories/infrastructure/services/ProductCaloriesService.scala (hunk 2 above)
Comment (Ledostuff) [reply src/main/scala/ru/ledostuff/calories/infrastructure/services/ProductCaloriesService.scala]: `
EitherT.liftF({engNamesRequests.traverse(e => e).map(_.collect({case Right(e) => e}))}).ensure(ProductCaloriesNotFound)(_.nonEmpty)
`
For me it looks the same. Could we make a filter in one go?

Comment (Phill101) [reply src/main/scala/ru/ledostuff/calories/infrastructure/services/ProductCaloriesService.scala]: I meant this:
```scala
val engNamesRequests = savedProduct.engNames.traverse { engName =>
  (EitherT.liftF[F, ErrorResponse, Unit](log.info(s"search calories for name $engName")) *>
    EitherT.fromOptionF[F, ErrorResponse, ProductCalories](productCaloriesService.getCaloriesByProductName(engName).value, ProductCaloriesNotFound)).value
}
EitherT.liftF(engNamesRequests.map(l => l.collect{case Right(e) => e})).ensure(ProductCaloriesNotFound)(_.nonEmpty)
```

We can simplify it:
```scala
val engNamesRequests = savedProduct.engNames.traverse { engName =>
  log.info(s"search calories for name $engName") *>
  productCaloriesService.getCaloriesByProductName(engName).value
}
EitherT.liftF(engNamesRequests.map(l => l.collect{case Some(e) => e})).ensure(ProductCaloriesNotFound)(_.nonEmpty)
```

And finally, you are able to use a special version of `traverse` - `traverseFilter`:
```scala
val engNamesRequests = savedProduct.engNames.traverseFilter { engName =>
  log.info(s"search calories for name $engName") *>
  productCaloriesService.getCaloriesByProductName(engName).value
}
EitherT.liftF(engNamesRequests).ensure(ProductCaloriesNotFound)(_.nonEmpty)
```

Comment (Phill101) [reply src/main/scala/ru/ledostuff/calories/infrastructure/services/ProductCaloriesService.scala]: Moreover, you have an opportunity to encode "non-emptiness" on a type-level.

```scala
val engNamesRequests = savedProduct.engNames.traverseFilter { engName =>
log.info(s"search calories for name $engName") *>
  productCaloriesService.getCaloriesByProductName(engName).value
}.map(NonEmptyList.fromList)
EitherT.fromOptionF(engNamesRequests, ProductCaloriesNotFound)
```

**Pull request merged**